File: test/override_null.py
import override_null_ext

foo = override_null_ext.foo()
foo.f()

class PyFoo(override_null_ext.foo):
    def __init__(self):
        super(override_null_ext.foo, self).__init__(self)

pyfoo = PyFoo()
try:
    thrown = False
    pyfoo.f()
except:
    thrown = True
assert thrown

class PyFooOverride(override_null_ext.foo):
    def __init__(self):
        super(override_null_ext.foo, self).__init__(self)

    def f(self):
        return None

pyfoo = PyFooOverride()
pyfoo.f()

File: test/override_null.cpp
// Copyright David Abrahams 2006. Distributed under the Boost
// Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
#include <boost/noncopyable.hpp>
#include <boost/python.hpp>

using namespace boost;

struct foo {
  virtual PyObject* f() {
    return NULL;
  }
};

struct foo_wrap : foo, python::wrapper<foo> {
  virtual PyObject* f() {
    python::override const o = get_override("f");
    if (o) {
      return o();
    } else {
      return foo::f();
    }
  }
};

BOOST_PYTHON_MODULE(override_null_ext) {
  python::class_<foo_wrap, noncopyable>("foo"
    , python::no_init)
    .def(python::init<>())
    .def("f", &foo_wrap::f);
}
